### PR TITLE
Add experimental support for descending sorting keys in MergeTree tables

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

This PR introduces a new setting `allow_experimental_reverse_key` that enables support for descending sort order in MergeTree sorting keys. This is useful for time series analysis, especially TopN queries. Example usage:
```sql
CREATE TABLE example
(
    time DateTime,
    key Int32,
    value String
) ENGINE = MergeTree
ORDER BY (time DESC, key);  -- Descending order for 'time' field
SETTINGS allow_experimental_reverse_key = 1;

SELECT * from example WHERE key = 'xxx' ORDER BY time DESC LIMIT 10;  -- This will apply more efficient ReadInOrder optimization instead of ReadInReverseOrder
```

Some preliminary perf test:
```sql
create table event_log_asc (dt DateTime, key UInt16) engine MergeTree order by dt;

insert into event_log_asc select now() - number, rand() from numbers(400000000);

optimize table event_log_asc final;

create table event_log_desc (dt DateTime, key UInt16) engine MergeTree order by dt desc settings allow_experimental_reverse_key = 1;

insert into event_log_desc select * from event_log_asc;

optimize table event_log_desc final;

SELECT *
FROM event_log_asc
WHERE key = 123
ORDER BY dt DESC
LIMIT 10

10 rows in set. Elapsed: 0.109 sec. Processed 27.53 million rows, 61.34 MB (253.50 million rows/s., 564.70 MB/s.)

SELECT *
FROM event_log_desc
WHERE key = 123
ORDER BY dt DESC
LIMIT 10

10 rows in set. Elapsed: 0.024 sec. Processed 25.44 million rows, 56.27 MB (1.06 billion rows/s., 2.35 GB/s.)
```

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit 4220b9002cc814db1db52f1f3921493dcae4fc45 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71095/4220b9002cc814db1db52f1f3921493dcae4fc45/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/c82cf25b3e5864bcc153cbe45adb8c6527e1ec6e/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeSettings.cpp [hunk 1]
```diff
@@ -202,6 +202,7 @@ namespace ErrorCodes
     DECLARE(String, storage_policy, "default", "Name of storage disk policy", 0) \
     DECLARE(String, disk, "", "Name of storage disk. Can be specified instead of storage policy.", 0) \
     DECLARE(Bool, allow_nullable_key, false, "Allow Nullable types as primary keys.", 0) \
+    DECLARE(Bool, allow_experimental_reverse_key, false, "Allow descending sorting key in MergeTree tables (experimental feature).", 0) \
```
[the remark this answers is not in the document]
Comment (orloffv) [reply src/Storages/MergeTree/MergeTreeSettings.cpp:205]: seems to need validation in the code

**Comment by alexey-milovidov** [comment]
It does not give any speedup:
https://pastila.nl/?0000bed8/597da07056e3b91a6420a54b22a852d9#kPeUfAAhFC62LfTR1Qaa2A==

**Comment by alexey-milovidov** [comment]
Although, there is good speedup when I read a small number of columns.

**Comment by amosbird** [comment]
> It does not give any speedup:

Are you using this query to verify?

```
SELECT *
FROM hits_desc
ORDER BY EventTime DESC
LIMIT 10
FORMAT Vertical
```

You can try adding non-primary key filters that cannot be satisfied by the last few granules alone (e.g., filters that result in fewer than 10 matches).

**Pull request merged**